broadinstitute/single_cell_portal_core
# Use serverState for detecting raw counts uploads and updating association selects (SCP-5221)

#### BACKGROUND
We require all new studies to supply raw counts files before they can add processed expression matrices.  However, the user must reload the page after uploading the file to populate in the values for raw counts files.  This is completely opaque to the user, and we have received recent help desk requests to this affect.

#### CHANGES
This update now dynamically updates the `Associated raw counts files` select menu for processed expression matrices after successful uploads.  Once the file pushes to the server and the `UploadWizard` component polls for new state, any additions/deletions to raw counts files will be present in the select menu.

#### MANUAL TESTING
1. Boot as normal, including DelayedJob and sign in
2. Create a new study and choose the "Classic" UX
3. Click into the `Processed matrices` tab and confirm that `Associated raw counts files` has no options
4. Click back to `Raw count matrices` and upload `test/test_data/expression_matrix_example.txt`
5. Wait at least 10 seconds to allow the file to fully push to the server and save, and for the UI to poll for updates
6. Go back to the `Processed matrices` tab and confirm that `Associated raw counts files` now has `expression_matrix_example.txt`
7. (Optional) Wait a few minutes for the ingest to complete, and then delete the raw count matrix
8. Wait another 10 seconds and go back to the `Processed matrices` tab and confirm that `Associated raw counts files` is empty again
